src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -1015,21 +1015,32 @@ bool is_array_superset(
         return false;
     }
 
-    // no mismatching elements. two possible cases
-    // 1. older_tuple_schema.Size() >= newer_tuple_schema.Size() ->
-    // compatible (implies newer_it==end())
-    // 2. older_tuple_schema.Size() <  newer_tuple_schema.Size() -> check
-    // excess elements with older["additionalItems"]
+    // no mismatching elements, and they don't have the same size.
+    // To be compatible, excess elements needs to be compatible with the other
+    // "additionalItems" schema
 
     auto older_additional_schema = get_object_or_empty(
       ctx.older, older, get_additional_items_kw(ctx.older.dialect()));
+    if (!std::all_of(
+          newer_it, newer_tuple_schema.end(), [&](json::Value const& n) {
+              return is_superset(ctx, older_additional_schema, n);
+          })) {
+        // newer has excess elements that are not compatible with
+        // older["additionalItems"]
+        return false;
+    }
 
-    // check that all excess schemas are compatible with
-    // older["additionalItems"]
-    return std::all_of(
-      newer_it, newer_tuple_schema.end(), [&](json::Value const& n) {
-          return is_superset(ctx, older_additional_schema, n);
-      });
+    auto newer_additional_schema = get_object_or_empty(
+      ctx.newer, newer, get_additional_items_kw(ctx.newer.dialect()));
+    if (!std::all_of(
+          older_it, older_tuple_schema.end(), [&](json::Value const& o) {
+              return is_superset(ctx, o, newer_additional_schema);
+          })) {
+        // older has excess elements that are not compatible with
+        // newer["additionalItems"]
+        return false;
```

Comment:
This commit makes the compact checks more restrictive but the new tests only cover the compatible cases and not these `return false;` cases. I think it would be useful to add 2 tests to cover these incompatible cases as well.

I think this is one example that results in an `ITEM_REMOVED_FROM_CLOSED_CONTENT_MODEL` error at the `BACKWARD` compatibility level:
```
{
  "type": "array",
  "items": [
    {
      "type": "boolean"
    },
    {
      "type": "integer"
    }
  ],
  "additionalItems": false
}
 ---->
{
  "type": "array",
  "items": [
    {
      "type": "boolean"
    }
  ]
}
```

Reply:
I added a negative test case and restricted another to show the rejection modes